src/column/reader.rs

```diff
@@ -164,10 +164,25 @@ impl<'a, T: DataType> ColumnReaderImpl<'a, T> where T: 'static {
       }
 
       // Batch size for the current iteration
-      let iter_batch_size = ::std::cmp::min(
-        batch_size,
-        (self.num_buffered_values - self.num_decoded_values) as usize
-      );
+      let iter_batch_size = {
+        // Compute approximate value based on values decoded so far
+        let mut adjusted_size = ::std::cmp::min(
+          batch_size,
+          (self.num_buffered_values - self.num_decoded_values) as usize
+        );
+
+        // Adjust batch size by taking into account how much space is left in values
+        // slice or levels slices (if available)
+        adjusted_size = ::std::cmp::min(adjusted_size, values.len() - values_read);
```

Comment:
seems this makes line 150 - 157 redundant - perhaps we can remove those (but might need to change `iter_batch_size` to `batch_size` or something similar).

Reply:
Those lines are still required, they adjust batch size as the smallest value between `batch_size`, `values`, `def_levels`, and `rep_levels`. Without it we might not exist the while loop if `batch_size` > `values_read` or `levels_read`, for example.

Reply:
You are right. That's why I think you need to change `let iter_batch_size = ...` to `batch_size = ...`, and make the parameter `batch_size` mutable.

Reply:
Sorry, I am not trying to avoid changing the code. I am more than happy to update it!

I was considering this initially, but then I thought that it would affect condition in the while loop, plus, since we decrease batch size with a series of `min(...)`, we would be inserting very small number of elements, even though we could insert full batch size. For example,

Page 1 has 3 values left, batch size is 30 (and we can buffer all 30 values), we would update batch size to be 3, as min(3, 30), read all those values, and then buffer page 2, which will have 50 values, but we would keep inserting by 3, rather than selecting 27, because of min(3, 50).

Let me know if my comment above does not make sense, or it is incorrect in terms of workflow of `read_batch`.

Reply:
Hmm you are right. Sorry for the wrong suggestion. Let's leave it then - we can think about better ways to remove this duplication in future.